core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -1342,8 +1344,28 @@ public UnifyResult apply(UnifyRuleCall call) {
       }
       final MutableAggregate aggregate2 =
           permute(query, project.getInput(), mapping.inverse());
-      final MutableRel result = unifyAggregates(aggregate2, target);
-      return result == null ? null : call.result(result);
+      final MutableRel unifiedAggregate = unifyAggregates(aggregate2, target);
+      if (unifiedAggregate == null) {
+        return null;
+      }
+
+      MutableRel result = unifiedAggregate;
+
+      //add Project if the mapping breaks order of fields in GroupSet
+      if (!Mappings.keepsOrdering(mapping)) {
+        final List<Integer> posList = new ArrayList<>();
+        final int fieldCount = aggregate2.rowType.getFieldCount();
+        for (int group: aggregate2.groupSet) {
+          if (mapping.inverse().getTargetOpt(group) != -1) {
+            posList.add(mapping.inverse().getTarget(group));
```

Comment:
Can you use a variable to save `mapping.inverse()` so that we don't need to do it twice?